nomos-services/utils/src/overwatch/status/utils.rs

```diff
@@ -0,0 +1,296 @@
+/// It can wait from 1..n services to be ready.
+///
+/// # Arguments
+///
+/// Requires `RuntimeServiceId` to be defined in the current scope.
+///
+/// * `$overwatch_handle`: A reference to the `OverwatchHandle` that manages the
+///   services.
+/// * `$timeout`: An optional `Duration` that specifies how long to wait for the
+///   services to be ready. If `None`, it will wait indefinitely.
+/// * `$( $service_type:ident ),+`: A list of service types that should be
+///   checked for readiness.
+///
+/// # Returns
+///
+/// - `Ok(())` if all specified services are ready within the timeout.
+/// - `Err(DynError)` if any of the specified services are not ready within the
+///   timeout.
+///
+/// # Example
+///
+/// ```rust,ignore
+/// use std::time::Duration;
+/// use overwatch::overwatch::OverwatchHandle;
+/// use services_utils::overwatch::status::wait_until_services_are_ready;
+///
+/// // The following types would be defined as part of your Overwatch runtime.
+/// struct RuntimeServiceId;
+/// struct ServiceA;
+/// struct ServiceB;
+/// struct ServiceC;
+///
+/// // Mock function to get an OverwatchHandle
+/// fn get_overwatch_handle() -> OverwatchHandle<RuntimeServiceId> {
+///     unimplemented!()
+/// }
+///
+/// fn main() {
+///     let overwatch_handle = get_overwatch_handle();
+///     wait_until_services_are_ready!(
+///        &overwatch_handle,
+///        Some(Duration::from_secs(10)),
+///        ServiceA,
+///        ServiceB,
+///        ServiceC
+///     );
+/// }
+/// ```
+///
+/// A complete example with working service implementations is available in this
+/// module's tests.
+#[macro_export]
+macro_rules! wait_until_services_are_ready {
+    ( $overwatch_handle:expr, $timeout:expr, $( $service_type:ty ),+ ) => {
+        let overwatch_handle: &::overwatch::overwatch::OverwatchHandle<RuntimeServiceId> = $overwatch_handle;
+        let timeout: Option<::std::time::Duration> = $timeout;
+
+        let mut non_ready_services: Vec<$crate::overwatch::status::ServiceStatusEntry<RuntimeServiceId>> = Vec::new();
+
+        $(
+            if let Err(service_status) = overwatch_handle
+                .status_watcher::<$service_type>()
+                .await
+                .wait_for(::overwatch::services::status::ServiceStatus::Ready, timeout)
+                .await
+            {
+                let service_id = <RuntimeServiceId as ::overwatch::services::AsServiceId<$service_type>>::SERVICE_ID;
+                let service_status_entry = $crate::overwatch::status::ServiceStatusEntry::<RuntimeServiceId>::from_overwatch(service_id, service_status);
+                non_ready_services.push(service_status_entry);
+            }
+        )+;
+
+        if !non_ready_services.is_empty() {
+            let error: $crate::overwatch::status::ServiceStatusEntriesError<RuntimeServiceId> = non_ready_services.into();
+            return Err(::overwatch::DynError::from(error));
+        }
+    };
+}
```

Comment:
Ok, so the macro works but I would change a few behaviors:

* I think you can create a future for each one and then join those futures, then you can collect them. You will be able to reuse some of the utilities you made I think.
* Not sure about the timeout, if at None it just waits forever then I guess there is no harm on having it.
* This should be in a specific scope (just rewrap everything), then return a `Result` from that scope so it can be used a method instead.

Reply:
| Not sure about the timeout, if at None it just waits forever then I guess there is no harm on having it.
As in, default to `None`?